### PR TITLE
1110 codeupdate- inbound and Concurrent update rebase

### PR DESCRIPTION
**Pull request by sagisin**
This PR contains the following commits:
1. Concurrent code update support.
2. Bootside mapping
3. Support for boot types attributes
4. All the defect fixes for inband and out of band updates in pldm. 

commit list:
66f7ac79 (HEAD -> 1110) Trigger Assemblelids method and handle response (#402)
d0397ba6 PLDM:Wait on the Software Updater (#296)
a5ee378a PLDM: Make boot type as Reipl during MPIPL (#270)
fa3d99a7 PLDM: Reset image activation after sending end update event (#259)
795b9cfe PLDM: Reset image activation match (#241)
9622648c PLDM: Minimise sending BIOS attr change event (#193)
66f7ac79 (HEAD -> 1110) Trigger Assemblelids method and handle response (#402)
d0397ba6 PLDM:Wait on the Software Updater (#296)
a5ee378a PLDM: Make boot type as Reipl during MPIPL (#270)
fa3d99a7 PLDM: Reset image activation after sending end update event (#259)
795b9cfe PLDM: Reset image activation match (#241)
9622648c PLDM: Minimise sending BIOS attr change event (#193)
56cd8dd1 oem: ibm: Set next boot side attribute During a rename, set the next boot side to Perm (same as current) so that in case of a failed update, PHYP knows to boot from the Perm side. The next boot side would be set to Temp on a successful update (via the Priority callback function).
b6fae78a oem: ibm: Side information enhancements (#132)
aea0bc0a PLDM: Remove old boot side setting (#190)
70c1d40b PLDM: Fix code update Rename
abfbe210 PLDM: APR changes in inband CU path (#178)
dca892f7 Pldm RR crash fix (#92)
cb246f9f pldm: Support bios attributes for boot types
3d8f9d9c PLDM: Setting the slot power state to Off (#62)
cd50161d PLDM: Support for LID type running in constructLidpath (#161)
c674f9ca PLDM: Add file type running (#156)
640271bc PLDM: Fix ContainerID for rename sensor (#130)
beec0379 PLDM: Next Boot Support (#123)
1e4220e4 PLDM: Include additional debug traces (#118)
04533600 PLDM bootside bump failure fix (#103)
bd85888e PLDM Fix for bootside attribute defect (#99)
44982b70 oem-ibm: Eliminate the Boot side Effecters and sensors
3e4c6d28 oem-ibm: State effecter for rename operation
bba274a4 oem-ibm: Maintain bootside Mapping and Set bios attribute
c1b78781 oem-ibm: support concurrent FW update

### CODE REVIEW

**Comment by manojkiraneda** [comment]
looking at the list , on a first glance i feel https://github.com/ibm-openbmc/pldm/commit/3d8f9d9c9b696b4eacfe78d77d32eeafba9b38ae PLDM: Setting the slot power state to Off (https://github.com/ibm-openbmc/pldm/pull/62)  is not required, its already merged upstream and downstream.

File: common/utils.cpp [hunk 1]
```diff
@@ -835,5 +835,77 @@ std::vector<pldm::pdr::EffecterID>
...
+        auto method = bus.new_method_call(
+            service.c_str(), biosConfigPath,
+            "xyz.openbmc_project.BIOSConfig.Manager", "GetAttribute");
+        method.append(dbusAttrName);
+        auto reply = bus.call(method);
+        reply.read(var1, var2, var3);
+    }
+    catch (const sdbusplus::exception::SdBusError& e)
+    {
+        std::cout << "Error getting the bios attribute"
```
Comment (Pavithrab7) [inline common/utils.cpp:861]: lg2 logs

File: common/utils.cpp [hunk 2]
```diff
@@ -835,5 +835,77 @@ std::vector<pldm::pdr::EffecterID>
...
+            auto method =
+                bus.new_method_call(service.c_str(), biosConfigPath,
+                                    SYSTEMD_PROPERTY_INTERFACE, "Set");
+            method.append(
+                biosConfigIntf, "PendingAttributes",
+                std::variant<PendingAttributesType>(pendingAttributes));
+            bus.call_noreply(method);
+        }
+        catch (const sdbusplus::exception::SdBusError& e)
+        {
```
Comment (Pavithrab7) [inline common/utils.cpp:901]: lg2

File: common/utils.cpp [hunk 3]
```diff
@@ -835,5 +835,77 @@ std::vector<pldm::pdr::EffecterID>
...
+
+void setBiosAttr(const BiosAttributeList& biosAttrList)
+{
+    static constexpr auto SYSTEMD_PROPERTY_INTERFACE =
+        "org.freedesktop.DBus.Properties";
+    constexpr auto biosConfigPath = "/xyz/openbmc_project/bios_config/manager";
+    constexpr auto biosConfigIntf = "xyz.openbmc_project.BIOSConfig.Manager";
+
+    constexpr auto dbusAttrType =
+        "xyz.openbmc_project.BIOSConfig.Manager.AttributeType.Enumeration";
```
Comment (Pavithrab7) [inline common/utils.cpp:878]: is this only used to set the enum attributes?

File: host-bmc/host_pdr_handler.cpp [hunk 4]
```diff
@@ -161,6 +166,14 @@ HostPDRHandler::HostPDRHandler(
                     pldm_entity obj{};
                     this->objPathMap[element.first] = obj;
                 }
+                else if (propVal ==
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.cpp:169]: this check is already there below can we have the `handleBootTypesAtPowerOn` called in that else if itself? this is a repeatitive else-if

File: host-bmc/host_pdr_handler.cpp [hunk 5]
```diff
@@ -176,6 +189,58 @@ HostPDRHandler::HostPDRHandler(
             }
         }
     });
+
+    chassisOffMatch = std::make_unique<sdbusplus::bus::match::match>(
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.cpp:193]: this off match was moved to oem-ibm  for power state changes below. Can we move this to oem-ibm?

File: libpldmresponder/bios_config.cpp [hunk 6]
```diff
@@ -523,6 +523,28 @@ void BIOSConfig::updateBaseBIOSTableProperty()
         auto method = bus.new_method_call(service.c_str(), biosConfigPath,
                                           dbusProperties, "Set");
         std::variant<BaseBIOSTable> value = baseBIOSTableMaps;
+#ifdef OEM_IBM
```
Comment (Pavithrab7) [inline libpldmresponder/bios_config.cpp:526]: can we void this is ifdef OEM_IBM?

File: oem/ibm/libpldmresponder/inband_code_update.cpp [hunk 7]
```diff
@@ -222,9 +313,13 @@ void CodeUpdate::setVersions()
                 auto imageInterface = "xyz.openbmc_project.Software.Activation";
                 auto imageObjPath = path.str.c_str();
 
-                try
-                {
-                    if (isCodeUpdateInProgress())
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/inband_code_update.cpp:227]: why is this check removed and added inside try? is there need to do a getDBUSVariant call if code update is not in progress?

File: oem/ibm/libpldmresponder/inband_code_update.cpp [hunk 8]
```diff
@@ -560,114 +750,33 @@ int processCodeUpdateLid(const std::string& filePath)
...
+        bus.call_noreply(method);
     }
-    else
+    catch (const std::exception& e)
     {
-        error("Error occurred during fork, error number - {ERROR_NUM}}",
-              "ERROR_NUM", errno);
-        return PLDM_ERROR;
+        error(
+            "InbandCodeUpdate: Failed to Assemble code update image ERROR={ERR_EXCEP}",
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/inband_code_update.cpp:771]: use `e` instead of e.what()

File: oem/ibm/libpldmresponder/inband_code_update.hpp [hunk 9]
```diff
@@ -13,12 +13,20 @@ namespace responder
 
 static constexpr uint8_t pSideNum = 1;
 static constexpr uint8_t tSideNum = 2;
+static constexpr uint8_t maxVersionRetry = 20;
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/inband_code_update.hpp:16]: can we add a comment telling why was this decided 20?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 10]
```diff
@@ -90,17 +90,12 @@ int pldm::responder::oem_ibm_platform::Handler::
...
-                                 codeUpdate);
-            }
-            else if (entityType == PLDM_OEM_IBM_ENTITY_FIRMWARE_UPDATE &&
-                     stateSetId == PLDM_OEM_IBM_FIRMWARE_UPDATE_STATE)
+                stateSetId == PLDM_OEM_IBM_FIRMWARE_UPDATE_STATE)
             {
                 if (stateField[currState].effecter_state ==
                     uint8_t(CodeUpdateState::START))
                 {
+                    std::cout << "Received Start Update Request From PHYP\n";
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:98]: info trace?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 11]
```diff
@@ -112,6 +107,7 @@ int pldm::responder::oem_ibm_platform::Handler::
                 else if (stateField[currState].effecter_state ==
                          uint8_t(CodeUpdateState::END))
                 {
+                    std::cout << "Received End Update Request From PHYP\n";
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:110]: info trace?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 12]
```diff
@@ -126,6 +122,7 @@ int pldm::responder::oem_ibm_platform::Handler::
                 else if (stateField[currState].effecter_state ==
                          uint8_t(CodeUpdateState::ABORT))
                 {
+                    std::cout << "Received Abort Update Request From PHYP\n";
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:125]: info trace?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 13]
```diff
@@ -794,6 +798,7 @@ void pldm::responder::oem_ibm_platform::Handler::_processEndUpdate(
     sdeventplus::source::EventBase& /*source */)
 {
     assembleImageEvent.reset();
+    std::cout << "Starting assembleCodeUpdateImage \n";
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:801]: need this?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 14]
```diff
@@ -817,6 +822,7 @@ void pldm::responder::oem_ibm_platform::Handler::_processStartUpdate(
         state = CodeUpdateState::FAIL;
     }
     auto sensorId = codeUpdate->getFirmwareUpdateSensor();
+    std::cout << "Sending Start Update sensor event to PHYP\n";
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:825]: this too?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 15]
```diff
@@ -850,13 +856,17 @@ void pldm::responder::oem_ibm_platform::Handler::updateOemDbusPaths(
...
+    biosAttrList.push_back(std::make_pair("pvm_boot_initiator", "Host"));
+    setBiosAttr(biosAttrList);
     pldm::utils::PropertyValue value =
         "xyz.openbmc_project.State.Chassis.Transition.Off";
     pldm::utils::DBusMapping dbusMapping{"/xyz/openbmc_project/state/chassis0",
                                          "xyz.openbmc_project.State.Chassis",
                                          "RequestedPowerTransition", "string"};
     try
     {
+        std::cout << "InbandCodeUpdate: ChassifOff the host\n";
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:869]: redundant?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 16]
```diff
@@ -891,28 +901,40 @@ void pldm::responder::oem_ibm_platform::Handler::_processSystemReboot(
...
+                            "Policy.AlwaysOn";
+                    try
+                    {
+                        info("InbandCodeUpdate: Setting the one time APR policy");
+                        dBusIntf->setDbusProperty(dbusMapping, value);
+                    }
+                    catch (const std::exception& e)
+                    {
+                        error(
+                        "Setting one-time restore policy failed, unable to set property PowerRestorePolicy ERROR={ERR_EXCEP}",
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:919]: use `e`

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 17]
```diff
@@ -1261,6 +1283,80 @@ void pldm::responder::oem_ibm_platform::Handler::modifyPDROemActions(
...
+        {
+            restartCause =
+                pldm::utils::DBusHandler().getDbusProperty<std::string>(
+                    "/xyz/openbmc_project/state/host0", "RestartCause",
+                    "xyz.openbmc_project.State.Host");
+            setBootTypesBiosAttr(restartCause);
+        }
+        catch (const std::exception& e)
+        {
+            std::cerr << "Failed to fetch the restartCause, ERROR=" << e.what()
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1304]: lg2

File: oem/ibm/libpldmresponder/oem_ibm_handler.hpp [hunk 18]
```diff
@@ -219,6 +236,78 @@ class Handler : public oem_platform::Handler
...
+            [this](sdbusplus::message::message& msg) {
+                sdbusplus::message::object_path path;
+                std::map<std::string, std::map<std::string, dbus::Value>>
+                    interfaces;
+                msg.read(path, interfaces);
+
+                for (auto interface : interfaces)
+                {
+                    std::cout << interface.first;
+                }
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.hpp:293]: redundant for loop?

File: oem/ibm/libpldmresponder/oem_ibm_handler.hpp [hunk 19]
```diff
@@ -493,6 +601,12 @@ class Handler : public oem_platform::Handler
 
     /** @brief Timer used for monitoring surveillance pings from host */
     sdeventplus::utility::Timer<sdeventplus::ClockId::Monotonic> timer;
+    /** @brief D-Bus Interfaced added signal match for Entity Manager */
+    std::unique_ptr<sdbusplus::bus::match::match> ibmCompatibleMatch;
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.hpp:605]: alias match_t in all the signal matches

File: softoff/softoff.cpp [hunk 20]
```diff
@@ -285,13 +285,27 @@ int SoftPowerOff::hostSoftOff(sdeventplus::Event& event)
...
+    uint8_t effecterState;
+    auto requestHostTransition =
+        pldm::utils::DBusHandler().getDbusProperty<std::string>(
+            "/xyz/openbmc_project/state/host0", "RequestedHostTransition",
+            "xyz.openbmc_project.State.Host");
+    if (requestHostTransition !=
+        "xyz.openbmc_project.State.Host.Transition.Off")
+    {
+        effecterState = PLDM_SW_TERM_GRACEFUL_RESTART_REQUESTED;
+    }
```
Comment (Pavithrab7) [inline softoff/softoff.cpp:297]: can we use a ternary operator here?

> effecterState = (requestHostTransition != "xyz.openbmc_project.State.Host.Transition.Off")
>     ? PLDM_SW_TERM_GRACEFUL_RESTART_REQUESTED
>     : PLDM_SW_TERM_GRACEFUL_SHUTDOWN_REQUESTED;

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 21]
```diff
@@ -1261,6 +1283,80 @@ void pldm::responder::oem_ibm_platform::Handler::modifyPDROemActions(
     }
 }
 
+void pldm::responder::oem_ibm_platform::Handler::handleBootTypesAtPowerOn()
+{
+    BiosAttributeList biosAttrList;
+    auto bootInitiator = getBiosAttrValue("pvm_boot_initiator_current");
+    std::string restartCause;
```
Comment (Pavithrab7) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1290]: this is redundant declaration? I see it declared in .hpp

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

File: common/utils.cpp [hunk 22]
```diff
@@ -835,5 +835,77 @@ std::vector<pldm::pdr::EffecterID>
...
+    auto& bus = DBusHandler::getBus();
+    try
+    {
+        auto service = pldm::utils::DBusHandler().getService(biosConfigPath,
+                                                             biosConfigIntf);
+        auto method = bus.new_method_call(
+            service.c_str(), biosConfigPath,
+            "xyz.openbmc_project.BIOSConfig.Manager", "GetAttribute");
+        method.append(dbusAttrName);
+        auto reply = bus.call(method);
```
Comment (manojkiraneda) [inline common/utils.cpp:856]: do you wanna block here for 25seconds here

File: host-bmc/host_pdr_handler.cpp [hunk 23]
```diff
@@ -113,13 +113,14 @@ HostPDRHandler::HostPDRHandler(
     oemUtilsHandler(oemUtilsHandler)
 {
     isHostOff = false;
-    isHostTransitioningToOff = false;
+    isHostRunning = false;
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.cpp:116]: this change is merged, you could do a rebase

File: host-bmc/host_pdr_handler.hpp [hunk 24]
```diff
@@ -456,8 +456,8 @@ class HostPDRHandler
 
     bool isHostOff;
 
-    /** @brief true/false based on the host transitioning value */
-    bool isHostTransitioningToOff;
+    /** @brief true/false based on the host is running or not */
+    bool isHostRunning;
```
Comment (Pavithrab7) [inline host-bmc/host_pdr_handler.hpp:460]: same as above this change is already merged, we can do a rebase

**Review by Pavithrab7 — CHANGES_REQUESTED** [review]

File: host-bmc/host_pdr_handler.cpp (hunk 23 above)
Comment (manojkiraneda) [inline host-bmc/host_pdr_handler.cpp:116]: this is already i think added by @Pavithrab7 , i am not sure why isHostTransitioningToOff is being removed ?

File: host-bmc/host_pdr_handler.cpp [hunk 25]
```diff
@@ -113,13 +113,14 @@ HostPDRHandler::HostPDRHandler(
...
     mergedHostParents = false;
     hostOffMatch = std::make_unique<sdbusplus::bus::match_t>(
         pldm::utils::DBusHandler::getBus(),
         propertiesChanged("/xyz/openbmc_project/state/host0",
                           "xyz.openbmc_project.State.Host"),
-        [this, repo, entityTree, bmcEntityTree](sdbusplus::message_t& msg) {
+        [this, repo, entityTree, bmcEntityTree,
+         oemPlatformHandler](sdbusplus::message::message& msg) {
```
Comment (manojkiraneda) [inline host-bmc/host_pdr_handler.cpp:123]: this should not be reverted from master, please change it back.

File: libpldmresponder/bios_config.cpp (hunk 6 above)
Comment (manojkiraneda) [reply libpldmresponder/bios_config.cpp:526]: +1

File: oem/ibm/libpldmresponder/file_io_by_type.cpp [hunk 26]
```diff
@@ -435,6 +435,11 @@ std::unique_ptr<FileHandler> getHandlerByType(uint16_t fileType,
             return std::make_unique<LidHandler>(fileHandle, false,
                                                 PLDM_FILE_TYPE_LID_MARKER);
         }
+        case PLDM_FILE_TYPE_LID_RUNNING:
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/file_io_by_type.cpp:438]: it is already there, below why is it removed and added back here.

File: oem/ibm/libpldmresponder/inband_code_update.cpp [hunk 27]
```diff
@@ -186,12 +211,82 @@ void CodeUpdate::setVersions()
...
     {
         error(
             "Failed to make a d-bus call to Object Mapper Association, error - {ERROR}",
             "ERROR", e);
+        if (retrySetVersion < maxVersionRetry)
+        {
+            retrySetVersion++;
+            usleep(500 * 1000);
+            setVersions();
+        }
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/inband_code_update.cpp:285]: do we still need this hack in pldm ?

File: oem/ibm/libpldmresponder/inband_code_update.cpp [hunk 28]
```diff
@@ -270,7 +371,7 @@ void CodeUpdate::setVersions()
                                     propertiesChanged(newImageId,
                                                       "xyz.openbmc_project."
                                                       "Software.Activation"),
-                                    [this](sdbusplus::message_t& msg) {
+                                    [this](sdbusplus::message::message& msg) {
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/inband_code_update.cpp:374]: dont revert the master change here.

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 29]
```diff
@@ -682,6 +679,9 @@ void pldm::responder::oem_ibm_platform::Handler::buildOEMPDR(
     realSAISensorId = findStateSensorId(
         repo.getPdr(), 0, PLDM_OEM_IBM_ENTITY_REAL_SAI, ENTITY_INSTANCE_1, 1,
         PLDM_STATE_SET_OPERATIONAL_FAULT_STATUS);
+    buildAllCodeUpdateSensorPDR(this, PLDM_OEM_IBM_ENTITY_FIRMWARE_UPDATE,
+                                ENTITY_INSTANCE_0,
+                                PLDM_OEM_IBM_BOOT_SIDE_RENAME, repo);
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:684]: why was this added twice ?

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 13 above)
Comment (manojkiraneda) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:801]: if so lg2

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp (hunk 14 above)
Comment (manojkiraneda) [reply oem/ibm/libpldmresponder/oem_ibm_handler.cpp:825]: lg2

File: oem/ibm/libpldmresponder/oem_ibm_handler.cpp [hunk 30]
```diff
@@ -1047,43 +1059,6 @@ void Handler::processSetEventReceiver()
...
-    else if (timer.isEnabled())
-    {
-        info(
-            "Failed to stop surveillance timer while remote terminus status is ‘{HOST_TRANST_OFF}’ with Terminus ID ‘{TID}’ ",
-            "HOST_TRANST_OFF", hostTransitioningToOff, "TID", tid);
-        startStopTimer(value);
-        pldm::utils::reportError(
-            "xyz.openbmc_project.PLDM.Error.setSurvTimer.RecvSurveillancePingFail");
-    }
-}
```
Comment (manojkiraneda) [inline oem/ibm/libpldmresponder/oem_ibm_handler.cpp:1085]: why was this removed ?

File: softoff/softoff.cpp (hunk 20 above)
Comment (manojkiraneda) [reply softoff/softoff.cpp:297]: why was soft off involved in code update ?

**Review by manojkiraneda — CHANGES_REQUESTED** [review]
